sklearn/preprocessing/tests/test_polynomial.py

```diff
@@ -890,6 +890,32 @@ def test_polynomial_features_deprecated_n_input_features():
         PolynomialFeatures().fit(X).n_input_features_
 
 
+def test_csr_polynomial_expansion_index_overflow():
+    N = 12
+    M = 120000
+    dtype = np.float32
+    x = sparse.csr_matrix(
+        (
+            np.arange(1, 5, dtype=np.int64),
+            (np.array([N - 1, N - 1, N, N]), np.array([M - 1, M, M - 1, M])),
+        ),
+        shape=(N + 1, M + 1),
+        dtype=dtype,
+        copy=False,
+    )
+    pf = PolynomialFeatures(interaction_only=True, include_bias=False, degree=2)
+    xinter = pf.fit_transform(x)
+    n_index, m_index = xinter.nonzero()
+
+    assert xinter.dtype == dtype
+    assert xinter.shape == (13, 7200180001)
+    assert_array_almost_equal(xinter.data, np.array([1, 2, 2, 3, 4, 12], dtype=dtype))
+    assert_array_almost_equal(n_index, np.array([11, 11, 11, 12, 12, 12]))
+    assert_array_almost_equal(
+        m_index, np.array([119999, 120000, 7200180000, 119999, 120000, 7200180000])
+    )
+
+
```

Comment:
Could you please write a test with `X.shape[1] == 65535` and `X.shape[1] == 65536` to check the dtype switch with `interaction_only=False`?

```
X = csr_matrix(([1.0], [65535 - 1], [0, 1]))
```


The limit comes from:

```python
>>> 1/2 * (np.sqrt(1 + 8 * np.iinfo(np.int32).max) - 1)
65535.49998664856
```

(I cheated with https://www.wolframalpha.com/input?i=%28d%5E2+%2B+d%29+%2F+2+%3D+e)

It could be relevant to do similar tests for `interaction_only=True` and `degree=3`.